### PR TITLE
Add FXIOS-7754 [v122] Add telemetry for SurfaceAdsImpression

### PR DESCRIPTION
**Pull request by PARAIPAN9**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-7754)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/17302)

## :bulb: Description

- Add glean metric
- Implement sending behavior of the event
## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [x] If needed I updated documentation / comments for complex code and public methods

### CODE REVIEW

**Comment by PARAIPAN9** [comment]
# Request for data collection review form

1) What questions will you answer with this data?
- How often are ads shown in the sidebar long enough (1.5 seconds) to count as an impression? For users who see ads, what else happens? Does displaying ads result in users not using the feature? Not using Firefox?

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? Some example responses:
- Business wants to show ads here, but also wants to understand potential impacts on feature retention and Firefox user retention.

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?
- This is the least invasive way of answering these questions--we don't track which ad is shown, just that an ad was shown.

4) Can current instrumentation answer these questions?
- No

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

<table>
  <tr>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>An ad was shown and visible in the sidebar for 1.5 seconds.</td>
    <td>Interaction</td>
    <td>#17302</td>
  </tr>
</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
 * Often the Privacy Notice for your product will link to where the documentation is expected to be.
 * Common examples for Mozilla products/services:
    * If this collection is Telemetry you can state "This collection is documented in its definitions files Histograms.json, Scalars.yaml, and/or Events.yaml and in the Probe Dictionary at https://probes.telemetry.mozilla.org."
    * If this data is collected using the Glean SDK you can state “This collection is documented in the Glean Dictionary at https://dictionary.telemetry.mozilla.org/"
 * In some cases, documentation is included in the project’s repository.

7) How long will this data be collected?  Choose one of the following:
- Currently set to expire 2024-06-01

8) What populations will you measure?
- All channels, all locales, all countries

9) If this data collection is default on, what is the opt-out mechanism for users?
- The user can opt out

10) Please provide a general description of how you will analyze this data.
- Glean

11) Where do you intend to share the results of your analysis?
- Glean and ad partners

12) Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection? If so:
- No third party tools

**Comment by PARAIPAN9** [comment]
@dnarcese && @perrymcmanis144

**Comment by dnarcese** [comment]
r+
# Data Review Form

_1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?_ 

Yes, telemetry is documented at https://dictionary.telemetry.mozilla.org/apps/firefox_ios

_2) Is there a control mechanism that allows the user to turn the data collection on and off?_ 

Yes, Firefox iOS provides data controls in the app settings

_3) If the request is for permanent data collection, is there someone who will monitor the data over time?_

This will be reviewed by me every 6 months

_4) Using the **[category system of data types](https://wiki.mozilla.org/Firefox/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?_

Type 2

_5) Is the data collection request for default-on or default-off?_

Default on

_6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?_

No

_7) Is the data collection covered by the existing Firefox privacy notice?_ 

Yes

_8) Does the data collection use a third-party collection tool?_ 

No

**Review by OrlaM — APPROVED** [review]

File: Client/Frontend/Fakespot/FakespotViewModel.swift [hunk 1]
```diff
@@ -467,6 +469,47 @@ class FakespotViewModel {
...
+
+    func handleVisibilityChanges(for view: UIView, in superview: UIView) {
+        let halfViewHeight = view.frame.height / 2
+        let intersection = superview.bounds.intersection(view.frame)
+        let areViewsIntersected = intersection.height >= halfViewHeight
+
+        if areViewsIntersected {
+            guard !isViewVisible else { return }
+            isViewVisible.toggle()
+            startTimer()
```
[the remark this answers is not in the document]
Comment (PARAIPAN9) [reply Client/Frontend/Fakespot/FakespotViewModel.swift:505]: The startTimer() method assists us in triggering the telemetry event after 1.5 seconds, and the guard statement ensures that startTimer() is not called a second time if the user scrolls and the view remains visible on the screen. Additionally, we set the repeats property of the timer to nil.

[the remark this answers is not in the document]
Comment (thatswinnie) [reply Client/Frontend/Fakespot/FakespotViewModel.swift:505]: As the timer is set to null when in `stopTimer`, the next time the view is moved and the ad is now visible a new timer will be created and the telemetry will be send again. I confirmed it with debugging. Per [this thread](https://mozilla.slack.com/archives/C05GL786AF8/p1700569688356289) the telemetry should only be send once.

[the remark this answers is not in the document]
Comment (PARAIPAN9) [reply Client/Frontend/Fakespot/FakespotViewModel.swift:505]: Yes, that is the way it works right now. I didn't know that we only want to send an event and stop. As far as I know, impressions occur every time we view the ad. But no problem, I'll update the PR to make sure the event is sent once.

**Review by thatswinnie — APPROVED** [review]

**Pull request merged**